### PR TITLE
Lazy Expressions

### PR DESCRIPTION
**Pull request by 0aix**
Preview. Lots of edge cases to consider. 

Checkstyle and tests are dead.

### CODE REVIEW

File: src/main/java/com/yahoo/bullet/parsing/Query.java [hunk 1]
```diff
@@ -27,9 +29,9 @@
 @Getter @Setter @Slf4j
 public class Query implements Configurable, Initializable {
     @Expose
-    private Projection projection;
+    private Map<String, LazyExpression> projection;
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/parsing/Query.java]: If we're inverting this map (field -> newName) => (expression -> name), then let's break the API and make projection => projections and let it be a list of this Projection object. That way we get an ordering too

Comment (akshaisarma) [reply src/main/java/com/yahoo/bullet/parsing/Query.java]: Also, why do we only initialize a Projection object in RunningQuery#initialize? Why don't we parse it right here using GSON?

File: src/main/java/com/yahoo/bullet/parsing/Query.java [hunk 2]
```diff
@@ -90,14 +95,14 @@ public void configure(BulletConfig config) {
     @Override
     public Optional<List<BulletError>> initialize() {
         List<BulletError> errors = new ArrayList<>();
-        if (filters != null) {
-            for (Clause clause : filters) {
-                clause.initialize().ifPresent(errors::addAll);
+        if (projection != null) {
+            if (projection.keySet().stream().anyMatch(s -> s.contains(DELIMITER))) {
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/parsing/Query.java]: This stuff should all be in Projection#initialize

File: src/main/java/com/yahoo/bullet/parsing/Query.java [hunk 3]
```diff
@@ -39,12 +41,16 @@
     @Expose
     private List<PostAggregation> postAggregations;
 
+    public static final BulletError NO_DELIMITERS = makeError("Projection fields cannot contain delimiters.",
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/parsing/Query.java]: Is this for BQL? I guess it can break aggregations etc too to have a field with a '.'

File: src/main/java/com/yahoo/bullet/parsing/Projection.java [hunk 4]
```diff
@@ -9,27 +9,27 @@
...
 public class Projection implements Configurable, Initializable {
-    /**
-     * The map of source field names to their new projected names.
-     */
     @Expose
-    private Map<String, String> fields;
+    private String name;
+
+    @Expose
+    private LazyExpression value;
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/parsing/Projection.java]: This can't be any LazyExpression right? Or are we allowing logical ands and stuff to appear here now?

File: src/main/java/com/yahoo/bullet/parsing/expressions/LazyExpression.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import static java.util.Arrays.asList;
+
+/**
+ * Use "Lazy" as a placeholder name because it rolls off the tongue.
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/parsing/expressions/LazyExpression.java]: .... You know what's better than things rolling off the tongue? Not rolling anything off the tongue. Just omit the Lazy... 🎉

File: src/main/java/com/yahoo/bullet/parsing/Query.java [hunk 6]
```diff
@@ -27,9 +29,9 @@
 @Getter @Setter @Slf4j
 public class Query implements Configurable, Initializable {
     @Expose
-    private Projection projection;
+    private Map<String, LazyExpression> projection;
     @Expose
-    private List<Clause> filters;
+    private LazyExpression filter;
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/parsing/Query.java]: Can we force this to be a LazyBinary?

File: src/main/java/com/yahoo/bullet/parsing/expressions/LazyExpression.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+ * - LazyValue
+ * - LazyField
+ * - LazyUnary
+ * - LazyBinary
+ * - LazyList
+ *
+ * LazyMap is not supported at the moment since there's a lot of type-safety to take into consideration.
+ *
+ * Also, note the "type" field in LazyExpression. This might be the true origin of the "lazy" in lazy expressions. The specified type
+ * is not a type-check; rather, when an expression is evaluated, it will force cast to that type (if specified) before returning, so it's like a
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/parsing/expressions/LazyExpression.java]: Beware the '

File: src/main/java/com/yahoo/bullet/parsing/expressions/LazyExpression.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+import static java.util.Arrays.asList;
+
+/**
+ * Use "Lazy" as a placeholder name because it rolls off the tongue.
+ *
+ * Lazy expressions are currently used in queries for filters and projections.
+ *
+ * A filter is simply a lazy expression. We accept a record iff it the expression evaluates to true (with an explicit cast to Boolean), e.g.
+ *
+ * "filter": { "left": {"field": "bcookie"}, "right": {"value": "123456", "type": "STRING"}, "op": "EQUALS" }
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/parsing/expressions/LazyExpression.java]: Use another value besides bcookie

File: src/main/java/com/yahoo/bullet/parsing/expressions/LazyExpression.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+ * really late type-check. Also, type must be primitive.
+ *
+ * Look at the Evaluator class to see how lazy expressions are evaluated.
+ */
+@Getter
+public abstract class LazyExpression implements Configurable, Initializable {
+    /** The type of the operation in binary/unary lazy expressions. */
+    @Getter
+    @AllArgsConstructor
+    public enum Operation {
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/parsing/expressions/LazyExpression.java]: Probably should be its own class

File: src/main/java/com/yahoo/bullet/parsing/expressions/LazyValue.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        value = null;
+        type = null;
+    }
+
+    @Override
+    public Optional<List<BulletError>> initialize() {
+        if (value == null) {
+            return Optional.of(Collections.singletonList(LAZY_VALUE_REQUIRES_NON_NULL_VALUE));
+        }
+        if (!Type.PRIMITIVES.contains(type)) {
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/parsing/expressions/LazyValue.java]: Eh, just assume to be String since it is...

File: src/main/java/com/yahoo/bullet/querying/evaluators/Evaluator.java [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    /**
+     * What is polymorphism?
+     *
+     * @param expression
+     * @return
+     */
+    public static Evaluator build(LazyExpression expression) {
+        if (expression instanceof LazyNull) {
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/querying/evaluators/Evaluator.java]: Just make each Expression return you the right Evaluator?

File: src/main/java/com/yahoo/bullet/querying/evaluators/BinaryOperations.java [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+
+    static BinaryOperator<TypedObject> OR = (left, right) -> {
+        return new TypedObject((Boolean) left.forceCast(Type.BOOLEAN).getValue() || (Boolean) right.forceCast(Type.BOOLEAN).getValue());
+    };
+
+    static BinaryOperator<TypedObject> XOR = (left, right) -> {
+        return new TypedObject((Boolean) left.forceCast(Type.BOOLEAN).getValue() ^ (Boolean) right.forceCast(Type.BOOLEAN).getValue());
+    };
+
+    static BinaryOperator<TypedObject> FILTER = (left, right) -> {
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/querying/evaluators/BinaryOperations.java]: How do you see this being used?

File: src/main/java/com/yahoo/bullet/querying/partitioning/SimpleEqualityPartitioner.java [hunk 13]
```diff
@@ -120,6 +121,8 @@ public SimpleEqualityPartitioner(BulletConfig config) {
         String key = fields.stream().map(fieldFilters::get).map(this::getFilterValue).collect(Collectors.joining(delimiter));
         // For the SimpleEqualityPartitioner, the query is mapped to exactly one key only.
         return Collections.singleton(key);
+        */
+        return Collections.singleton("");
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/querying/partitioning/SimpleEqualityPartitioner.java]: Uhh, gotta fix this

File: src/main/java/com/yahoo/bullet/parsing/Query.java (hunk 1 above)
Comment (akshaisarma) [reply src/main/java/com/yahoo/bullet/parsing/Query.java]: Ok ordering is not as important because we can't easily produce output in the same order (JSON). We should still use parsing.Projection here though.

File: src/main/java/com/yahoo/bullet/parsing/expressions/LazyExpression.java (hunk 8 above)
Comment (akshaisarma) [reply src/main/java/com/yahoo/bullet/parsing/expressions/LazyExpression.java]: Actually it's because it's open source that we don't want our internal fields out here

File: src/main/java/com/yahoo/bullet/parsing/expressions/LazyValue.java (hunk 10 above)
Comment (akshaisarma) [reply src/main/java/com/yahoo/bullet/parsing/expressions/LazyValue.java]: I know but if type isn't present, just assume String because value actually is still a string.

File: src/main/java/com/yahoo/bullet/querying/RunningQuery.java [hunk 14]
```diff
@@ -57,7 +60,12 @@ public RunningQuery(String id, String queryString, BulletConfig config) {
     @Override
     public Optional<List<BulletError>> initialize() {
         start();
-        return query.initialize();
+        Optional<List<BulletError>> errors = query.initialize();
+        if (!errors.isPresent()) {
+            filter = new Filter(query.getFilter());
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/querying/RunningQuery.java]: It's weird that we only do this for the new expression stuff. Aggs and post aggs are in Querier. I feel like it should all be in Querier.

File: src/main/java/com/yahoo/bullet/querying/evaluators/Evaluator.java (hunk 11 above)
Comment (akshaisarma) [reply src/main/java/com/yahoo/bullet/querying/evaluators/Evaluator.java]: I thought we agreed to return the right class as a compromise, since at least the logic isn't there and we don't have to get the Java police on us for breaking polymorphism

File: src/main/java/com/yahoo/bullet/parsing/Projection.java (hunk 4 above)
Comment (0aix) [reply src/main/java/com/yahoo/bullet/parsing/Projection.java]: I'm not sure why we would want to restrict logical operations. We can have boolean fields and allowing logical operations wouldn't break anything.

File: src/main/java/com/yahoo/bullet/parsing/Query.java (hunk 6 above)
Comment (akshaisarma) [reply src/main/java/com/yahoo/bullet/parsing/Query.java]: Sure, if you can you make sure we can only put expressions that produce booleans in here without force casting it?

File: src/main/java/com/yahoo/bullet/parsing/Query.java (hunk 1 above)
Comment (0aix) [reply src/main/java/com/yahoo/bullet/parsing/Query.java]: I could make it a List<Projection> projections; and in initialize(), check that all the names are unique. If you want a map in the projection, we could do some interesting things.

"projection": {
    "fields": {
    },
    "post": {
    }
}

File: src/main/java/com/yahoo/bullet/parsing/expressions/LazyExpression.java (hunk 8 above)
Comment (0aix) [reply src/main/java/com/yahoo/bullet/parsing/expressions/LazyExpression.java]: That's what I meant .-.

File: src/main/java/com/yahoo/bullet/parsing/expressions/LazyValue.java (hunk 10 above)
Comment (0aix) [reply src/main/java/com/yahoo/bullet/parsing/expressions/LazyValue.java]: ok fair

File: src/main/java/com/yahoo/bullet/querying/RunningQuery.java (hunk 14 above)
Comment (0aix) [reply src/main/java/com/yahoo/bullet/querying/RunningQuery.java]: Yea, that might work better. In Querier#initialize(), I can have 

filter = new Filter(runningQuery.getQuery().getFilter()); 
projection = new Projection(runningQuery.getQuery().getProjection());

after runningQuery#initialize()

File: src/main/java/com/yahoo/bullet/querying/evaluators/Evaluator.java (hunk 11 above)
Comment (0aix) [reply src/main/java/com/yahoo/bullet/querying/evaluators/Evaluator.java]: ok, I have no problems with that.

File: src/main/java/com/yahoo/bullet/parsing/expressions/LazyExpression.java (hunk 8 above)
Comment (akshaisarma) [reply src/main/java/com/yahoo/bullet/parsing/expressions/LazyExpression.java]: Oh I read your response as a statement not a question.
